ethereum/event/BREventQueue.c

```diff
@@ -129,56 +146,105 @@ eventQueueEnqueue (BREventQueue queue,
         queue->pending = this;
     }
 
+    if (signal) pthread_cond_signal (&queue->cond);
     pthread_mutex_unlock(&queue->lock);
 }
 
 extern void
 eventQueueEnqueueTail (BREventQueue queue,
                        const BREvent *event) {
-    eventQueueEnqueue (queue, event, 1);
+    eventQueueEnqueue (queue, event, 1, 0);
 }
 
 extern void
 eventQueueEnqueueHead (BREventQueue queue,
                        const BREvent *event) {
-    eventQueueEnqueue (queue, event, 0);
+    eventQueueEnqueue (queue, event, 0, 0);
+}
+extern void
+eventQueueEnqueueTailSignal (BREventQueue queue,
+                             const BREvent *event) {
+    eventQueueEnqueue (queue, event, 1, 1);
+}
+
+extern void
+eventQueueEnqueueHeadSignal (BREventQueue queue,
+                             const BREvent *event) {
+    eventQueueEnqueue (queue, event, 0, 1);
+}
+
+static int
+_eventQueueDequeue (BREventQueue queue,
+                    BREvent *event) {
+    // Get the next pending event
+    BREvent *this = queue->pending;
+
+    // if there is one, process it
+    if (NULL == this) return 0;
+
+    // Remove `this` from the pending list.
+    queue->pending = this->next;
+
+    // Fill in the provided event;
+    this->next = NULL;
+    memcpy (event, this, queue->size);
+
+    // Return `this` to the available list.
+    this->next = queue->available;
+    queue->available = this;
+
+    return 1;
 }
 
 extern BREventStatus
 eventQueueDequeue (BREventQueue queue,
                    BREvent *event) {
-    BREventStatus status = EVENT_STATUS_SUCCESS;
-    
+   if (NULL == event)
+        return EVENT_STATUS_NULL_EVENT;
+
+    pthread_mutex_lock (&queue->lock);
+    BREventStatus status = (_eventQueueDequeue (queue, event)
+                            ? EVENT_STATUS_SUCCESS
+                            : EVENT_STATUS_NONE_PENDING);
+    pthread_mutex_unlock(&queue->lock);
+
+    return status;
+}
+
+extern BREventStatus
+eventQueueDequeueWait (BREventQueue queue,
+                       BREvent *event) {
     if (NULL == event)
         return EVENT_STATUS_NULL_EVENT;
-    
-    pthread_mutex_lock(&queue->lock);
-    
-    // Get the next pending event
-    BREvent *this = queue->pending;
-    
-    // there is none, just return
-    if (NULL == this)
-        status = EVENT_STATUS_NONE_PENDING;
-    
-    // otherwise process it.
-    else {
-        // Remove `this` from the pending list.
-        queue->pending = this->next;
-        
-        // Fill in the provided event;
-        this->next = NULL;
-        memcpy (event, this, queue->size);
-        
-        // Return `this` to the available list.
-        this->next = queue->available;
-        queue->available = this;
-    }
+
+    BREventStatus status = EVENT_STATUS_SUCCESS;
+
+    pthread_mutex_lock (&queue->lock);
+    while (!queue->abort && !_eventQueueDequeue (queue, event))
+        if (0 != pthread_cond_wait (&queue->cond, &queue->lock))
+            return EVENT_STATUS_WAIT_ERROR;
```

Comment:
This would represent a likely catastrophic failure so probably moot but early return here without releasing the `lock`.

Reply:
Fixed, regarding a proper return path.